sdks/python/apache_beam/transforms/core.py

```diff
@@ -1544,10 +1544,16 @@ def Filter(fn, *args, **kwargs):  # pylint: disable=invalid-name
   # TODO: What about callable classes?
   if hasattr(fn, '__name__'):
     wrapper.__name__ = fn.__name__
+
+  # Get type hints from this instance or the callable. Do not use output type
+  # hints from the callable (which should be bool if set).
+  fn_type_hints = typehints.decorators.IOTypeHints.from_callable(fn)
+  if fn_type_hints is not None:
+    fn_type_hints.output_types = None
```

Comment:
With this change, do we still need both branches in line 1559, 1563? Perhaps we can make the evaluation more deterministic as in:
```
  if (get_type_hints(wrapper).input_types
      and get_type_hints(wrapper).input_types[0]):
    output_hint = get_type_hints(wrapper).input_types[0][0]
    get_type_hints(wrapper).set_output_types(typehints.Iterable[output_hint])
```

Reply:
After this PR, output_hint is set from one of (in order):
- fn._type_hints.output_types[0][0] (could be set using the with_output_types() decorator)
- fn._type_hints.input_types[0][0]
- IOTypeHints.from_callable(fn).input_types[0][0] (Py3 type hint syntax)

This change removes IOTypeHints.from_callable(fn).output_types[0][0] from the above list.

To answer your question, the branches are still required. The branch in line 1559 checks if output_hint is None and tries using input_types as a substitute. But output_hint may still remain None so line 1563 avoids setting an output type of Iterable[None].